File: junit5/src/test/java/BombSpec.java
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;

import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

public class BombSpec {

    @Test
    public void Bomb_tick_after_creating_it() {
        Bomb bomb = new Bomb();
        assertThat(bomb.state(), is("ticktock"));
    }

    @Test
    public void Bomb_disarmed_when_wires_are_cut_in_following_order_green_yellow_and_blue() {
        Bomb bomb = new Bomb();
        bomb.cutWires(Wire.green, Wire.yellow, Wire.blue);
        assertThat(bomb.state(), is("disarmed"));
    }

    @Nested
    public class Bomb_explode {

        @Test
        public void after_two_seconds() {
            Bomb bomb = new Bomb();
            waitOverTwoSeconds();
            assertThat(bomb.state(), is("kaboom!"));
        }

        @Test
        public void when_yellow_wire_is_cut_first() {
            Bomb bomb = new Bomb();
            bomb.cutWires(Wire.yellow);
            assertThat(bomb.state(), is("kaboom!"));
        }

        @Test
        public void when_blue_wire_is_cut_first() {
            Bomb bomb = new Bomb();
            bomb.cutWires(Wire.blue);
            assertThat(bomb.state(), is("kaboom!"));
        }

        @Test
        public void when_green_wire_is_cut_first_and_then_blue() {
            Bomb bomb = new Bomb();
            bomb.cutWires(Wire.blue, Wire.blue);
            assertThat(bomb.state(), is("kaboom!"));
        }
    }

    @Nested
    public class After_bomb_is_disarmed {

        @Test
        public void it_wont_explode_after_two_seconds() {
            Bomb bomb = new Bomb();
            bomb.cutWires(Wire.green, Wire.yellow, Wire.blue);
            assertThat(bomb.state(), is("disarmed"));
            waitOverTwoSeconds();
            assertThat(bomb.state(), is("disarmed"));
        }

        @Test
        public void it_wont_hurt_if_wires_are_cut_again() {
            Bomb bomb = new Bomb();
            bomb.cutWires(Wire.green, Wire.yellow, Wire.blue);
            assertThat(bomb.state(), is("disarmed"));
            bomb.cutWires(Wire.yellow);
            assertThat(bomb.state(), is("disarmed"));
        }

    }

    private void waitOverTwoSeconds() {
        try {
            Thread.sleep(2100);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }
}

File: kotlintest/src/test/kotlin/BombSpec.kt
import io.kotlintest.eventually
import io.kotlintest.matchers.fail
import io.kotlintest.matchers.should
import io.kotlintest.matchers.shouldBe
import io.kotlintest.matchers.shouldThrow
import io.kotlintest.specs.FreeSpec
import java.util.concurrent.TimeUnit

class BombSpec : FreeSpec() {

    init {

        """
Bomb will be ticking after it is initialized.
Bomb explode after two seconds.
Bomb is disposed if all wires are cut.
Bomb explode if wires are NOT cut in following order: green, yellow and blue
"""

        val waitOverTwoSeconds = fun() { Thread.sleep(2100) }

        "Bomb tick after creating it" {
            val bomb = Bomb()
            bomb.state() shouldBe "ticktock"
        }

        "Bomb explode" - {

            "after two seconds" {
                val bomb = Bomb()
                waitOverTwoSeconds()
                bomb.state() shouldBe "kaboom!"
            }

            "when yellow wire is cut first" {
                val bomb = Bomb()
                bomb.cutWires(Wire.yellow)
                bomb.state() shouldBe "kaboom!"
            }

            "when blue wire is cut first" {
                val bomb = Bomb()
                bomb.cutWires(Wire.blue)
                bomb.state() shouldBe "kaboom!"
            }

            "when green wire is cut first and then blue" {
                val bomb = Bomb()
                bomb.cutWires(Wire.green, Wire.blue)
                bomb.state() shouldBe "kaboom!"
            }

        }

        "Bomb disarmed when wires are cut in following order: green, yellow and blue" {
            val bomb = Bomb()
            bomb.cutWires(Wire.green, Wire.yellow, Wire.blue)
            bomb.state() shouldBe "disarmed"
        }

        "After bomb is disarmed" - {

            "it won't hurt if wires are cut again" {
                val bomb = Bomb()
                bomb.cutWires(Wire.green, Wire.yellow, Wire.blue)
                bomb.state() shouldBe "disarmed"
                bomb.cutWires(Wire.green)
            }

            "it wont explode after two seconds" {
                val bomb = Bomb()
                bomb.cutWires(Wire.green, Wire.yellow, Wire.blue)
                bomb.state() shouldBe "disarmed"
                waitOverTwoSeconds
                bomb.state() shouldBe "disarmed"
            }
        }

    }
}